File: app/assets/app/search-management.model.ts
export class SolrIndex {
  id: string;
  name: string;
  description: string;
}

export class SynonymRule {
  id: string;
  synonymType: number;
  term: string;
  isActive: boolean;
}

export class UpDownRule {
  id: string;
  upDownType?: number;
  boostMalusValue?: number;
  term: string;
  upDownDropdownDefinitionMapping?: number;
  suggestedSolrFieldName?: string;
  isActive: boolean;
}

// TODO consider other persistence solution (e.g. REST)
export const upDownDropdownDefinitionMappings = [
  { displayName: 'UP(+++++)', upDownType: 0, boostMalusValue: 500 },
  { displayName: 'UP(++++)', upDownType: 0, boostMalusValue: 100 },
  { displayName: 'UP(+++)', upDownType: 0, boostMalusValue: 50 },
  { displayName: 'UP(++)', upDownType: 0, boostMalusValue: 10 },
  { displayName: 'UP(+)', upDownType: 0, boostMalusValue: 5 },
  { displayName: 'DOWN(-)', upDownType: 1, boostMalusValue: 5 },
  { displayName: 'DOWN(--)', upDownType: 1, boostMalusValue: 10 },
  { displayName: 'DOWN(---)', upDownType: 1, boostMalusValue: 50 },
  { displayName: 'DOWN(----)', upDownType: 1, boostMalusValue: 100 },
  { displayName: 'DOWN(-----)', upDownType: 1, boostMalusValue: 500 }
];

export class FilterRule {
  id: string;
  term: string;
  suggestedSolrFieldName?: string;
  isActive: boolean;
}

export class DeleteRule {
  id: string;
  term: string;
  isActive: boolean;
}

export class RedirectRule {
  id: string;
  target: string;
  isActive: boolean;
}

export class SearchInput {
  id: string;
  term: string;
  synonymRules: Array<SynonymRule>;
  upDownRules: Array<UpDownRule>;
  filterRules: Array<FilterRule>;
  deleteRules: Array<DeleteRule>;
  redirectRules: Array<RedirectRule>;
}

export class SuggestedSolrField {
  id: string;
  name: string;
}

File: app/models/SearchManagementModel.scala
package models

package object SearchManagementModel {

  // TODO evaluate model adjustment for SolrIndex being leading construct and containts a List[SearchInput]
  case class SolrIndex(id: Option[String] = None,
                       name: String,
                       description: String)

  case class SynonymRule(id: Option[String] = None,
                         synonymType: Int,
                         term: String,
                         isActive: Boolean)

  case class UpDownRule(id: Option[String] = None,
                        upDownType: Int,
                        boostMalusValue: Int,
                        term: String,
                        isActive: Boolean)

  case class FilterRule(id: Option[String] = None,
                        term: String,
                        isActive: Boolean)

  case class DeleteRule(id: Option[String] = None,
                        term: String,
                        isActive: Boolean)

  // TODO rearrange SearchManagementRepository algorithms purely functional, so that no mutable var-attributes necessary
  case class SearchInput(id: Option[String] = None,
                         term: String,
                         var synonymRules: List[SynonymRule] = List.empty,
                         var upDownRules: List[UpDownRule] = List.empty,
                         var filterRules: List[FilterRule] = List.empty,
                         var deleteRules: List[DeleteRule] = List.empty)

  // TODO currently not in use ...
  // TODO Consider resolving as Option field in the Rule's itself (e.g. UpDownRule.errorMsgs List[String])
  case class SearchInputValidationResult(inputTermErrorMsg: Option[String],
                                         synonymRulesErrorMsg: List[Map[Long, String]],
                                         upDownRulesErrorMsg: List[Map[Long, String]],
                                         filterRulesErrorMsg: List[Map[Long, String]],
                                         deleteRulesErrorMsg: List[Map[Long, String]])

  case class SuggestedSolrField(id: Option[String] = None,
                                name: String)

}
